iree/samples/simple_embedding/simple_embedding_run_local_sync.c

```diff
@@ -0,0 +1,211 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// A example of setting up the HAL mddule to run simple pointwise array
+// multiplication with the dylib-sync driver.
+#include <stdio.h>
+
+#include "iree/base/api.h"
+#include "iree/hal/api.h"
+#include "iree/hal/local/executable_loader.h"
+#include "iree/hal/local/loaders/legacy_library_loader.h"
+#include "iree/hal/local/sync_driver.h"
+#include "iree/modules/hal/hal_module.h"
+#include "iree/vm/api.h"
+#include "iree/vm/bytecode_module.h"
+
+// Compiled module embedded here to avoid file IO:
+#if IREE_ARCH_RISCV_64
+#include "iree/samples/simple_embedding/simple_embedding_test_llvm_aot_rv64.h"
+#else
+#include "iree/samples/simple_embedding/simple_embedding_test_bytecode_module_c.h"
+#endif
+
+iree_status_t Run() {
+  // TODO(benvanik): move to instance-based registration.
+  IREE_RETURN_IF_ERROR(iree_hal_module_register_types());
+
+  iree_vm_instance_t* instance = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_vm_instance_create(iree_allocator_system(), &instance));
+
+  // Set paramters for the device created in the next ste.
+  iree_hal_sync_device_params_t params;
+  iree_hal_sync_device_params_initialize(&params);
+
+  iree_hal_executable_loader_t* dylib_loader = NULL;
+  // TODO(marbre): Use embedded instead of legacy loader.
+  iree_status_t status = iree_hal_legacy_library_loader_create(
+      iree_allocator_system(), &dylib_loader);
+  iree_hal_executable_loader_t* loaders[1] = {dylib_loader};
+  iree_hal_executable_loader_retain(loaders[0]);
+
+  iree_host_size_t loader_count = IREE_ARRAYSIZE(loaders);
+
+  iree_string_view_t identifier = iree_make_cstring_view("dylib");
+
+  // Create the synchronous device.
+  iree_hal_device_t* device = NULL;
+
+  IREE_RETURN_IF_ERROR(
+      iree_hal_sync_device_create(identifier, &params, loader_count, loaders,
+                                  iree_allocator_system(), &device));
+
+  // Create the corresponding HAL module.
+  iree_vm_module_t* hal_module = NULL;
+  IREE_RETURN_IF_ERROR(
+      iree_hal_module_create(device, iree_allocator_system(), &hal_module));
+
+  // Load bytecode module from the embedded data.
+#if IREE_ARCH_RISCV_64
+  const struct iree_file_toc_t* module_file_toc =
+      simple_embedding_test_llvm_aot_rv64_create();
+#else
+  // Note the setup here only supports native build. The bytecode is not built
+  // for the cross-compile execution. The code can be compiled but it will
+  // hit runtime error in a cross-compile environment.
+  const struct iree_file_toc_t* module_file_toc =
+      simple_embedding_test_bytecode_module_c_create();
+#endif
+
+  iree_vm_module_t* bytecode_module = NULL;
+  iree_const_byte_span_t module_data = {(const uint8_t*)module_file_toc->data,
```

Comment:
if you use `iree_make_const_byte_span` you won't need the cast - just `module_data = iree_make_const_byte_span(module_file_toc->data, module_file_toc->size);`